be/src/util/once.h

```diff
@@ -1,78 +1,54 @@
-//   http://www.apache.org/licenses/LICENSE-2.0
-//
-// Unless required by applicable law or agreed to in writing,
-// software distributed under the License is distributed on an
-// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
-// KIND, either express or implied.  See the License for the
-// specific language governing permissions and limitations
-// under the License.
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Limited.
 
 #pragma once
 
+#include <bthread/sys_futex.h>
+
 #include <atomic>
+#include <functional>
 
-#include "storage/olap_common.h"
+#include "common/statusor.h"
 
 namespace starrocks {
 
-// Utility class for implementing thread-safe call-once semantics.
-//
-// call() will return stored result regardless of whether the first invocation
-// returns a success status or not.
-//
-// Example:
-//   class Resource {
-//   public:
-//     Status init() {
-//       _init_once.call([this] { return _do_init(); });
-//     }
-//
-//     bool is_inited() const {
-//       return _init_once.has_called() && _init_once.stored_result().ok();
-//     }
-//   private:
-//     Status _do_init() { /* init logic here */ }
-//     StarRocksCallOnce<Status> _init_once;
-//   };
-template <typename ReturnType>
-class StarRocksCallOnce {
+class SuccessOnce {
 public:
-    StarRocksCallOnce() : _has_called(false) {}
-
-    // If the underlying `once_flag` has yet to be invoked, invokes the provided
-    // lambda and stores its return value. Otherwise, returns the stored Status.
-    template <typename Fn>
-    ReturnType call(Fn fn) {
-        std::call_once(_once_flag, [this, fn] {
-            _status = fn();
-            _has_called.store(true, std::memory_order_release);
-        });
-        return _status;
-    }
+    SuccessOnce() : _flag(0) {}
 
-    // Return whether `call` has been invoked or not.
-    bool has_called() const {
-        // std::memory_order_acquire here and std::memory_order_release in
-        // init(), taken together, mean that threads can safely synchronize on
-        // _has_called.
-        return _has_called.load(std::memory_order_acquire);
-    }
+    // Return value:
+    //  true: |fn| is invoked and returned an OK status.
+    //  false: |fn| has been invoked before and returned an OK status.
+    //  non-OK status: |fn| is invoked and returned a non-OK status.
+    StatusOr<bool> call(const std::function<Status()>& fn);
 
-    // Return the stored result. The result is only meaningful when `has_called() == true`.
-    ReturnType stored_result() const { return _status; }
+    bool has_succeeded() const { return _flag.load(std::memory_order_acquire) == 2; }
 
 private:
-    std::atomic<bool> _has_called;
-    std::once_flag _once_flag;
-    ReturnType _status;
+    std::atomic<int> _flag;
 };
 
+inline StatusOr<bool> SuccessOnce::call(const std::function<Status()>& fn) {
+    while (true) {
+        auto curr_flag = _flag.load(std::memory_order_acquire);
```

Comment:
Why not use std::call_once directly

Reply:
Updated, please check the comments above the function `success_once`.